### ISSUE TITLE
Computing loss outside model

### ISSUE THREAD

**Issue by abhaymittal** [facebookresearch/detectron2#1759]
## ❓ How to do something using detectron2
Hi, I am experimenting with detectron2 for my project and need to write a custom loss function. In the examples, I saw that the loss is usually made part of the model which is returned when the model is in training mode. However, what I'd like to do is to just return the outputs from the model (which I know how to) and then compute the loss outside in the main loop. I am stuck in the latter part. 

The main motivation for computing loss outside is that I need to pass an additional mask which is not part of the batch and keeps on changing during the execution of the program. 

- How do I call the loss function outside in a multi-gpu environment? I need to make sure that the masks for correct samples are sent to each gpu. 

- If the above is not recommended, is it possible to add a sample and class mask to the batch dicts itself which changes during the execution of the program? From what I understand, this cannot be a part of the DatasetMapper and loader.

Describe what you want to do, including:
1. what inputs you will provide, if any: COCO dataset
2. what outputs you are expecting: classification logits

Edit: Adding a bit more information to make it clearer. I have modified the dataset mapper and loader such that for each sample in batch, I get the sample ID. Now I have weights associated with each sample ID which change througghout the program. I want to compute the weighted loss and therefore, am looking for a way to either compute loss outside the model in a multi gpu environment or send in the sample weights correctly for each batch across gpus.

**Comment by ppwwyyxx** [comment]
If you'd like to do anything different from the standard main loop we have, it should be easy to just modify [the example](https://github.com/facebookresearch/detectron2/blob/b88c6c06563e4db1139aafbd6d8d97d1fa7a57e4/tools/plain_train_net.py#L161-L164).

However, since already need to customize the model, it might be easier to still compute losses in the model. It's possible to add extra inputs (e.g. "weights" in your issue) to the model by https://detectron2.readthedocs.io/tutorials/data_loading.html

If adding weights to losses is the only thing needed to do, it's easy to just write:
```
class NewModel(nn.Module):
    def __init__(...):
         self.old_model = build_model(...)
    def forward(...):
         losses = self.old_model.forward(...)
         return weighted_losses from losses
```
and use this NewModel for training

**Comment by abhaymittal** [comment]
Thanks for replying! In my case, the weights change during the execution of the program (active learning).  Therefore, as per my understanding, I can not directly add them to the data loader. One alternative that I have found is to build a custom dataset loader which also returns the sample idx (based on dataset dict) for each sample in  the batch and then I can manually add the weights in the batch after reading the batch. Is there another recommended way to do this?

**Comment by ppwwyyxx** [comment]
As https://detectron2.readthedocs.io/tutorials/data_loading.html shows, a custom data loader can do anything that python can do:

> The data loader is simply a python iterator that produces the format that the model accepts. You can implement it using any tools you like.

**Comment by abhaymittal** [comment]
Yep, I understand that. I was just wondering if there was more of a recommended way to tackle this. I'll close this issue.
